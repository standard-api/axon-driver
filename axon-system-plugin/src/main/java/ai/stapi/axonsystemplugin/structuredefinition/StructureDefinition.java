package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.axonsystem.graphaggregate.AggregateWithDynamicGraph;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.FixedImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.schema.structuredefinition.StructureDefinitionNormalizer;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.CreationPolicy;
import org.axonframework.spring.stereotype.Aggregate;
import org.springframework.beans.factory.annotation.Autowired;

@Aggregate
public class StructureDefinition extends AggregateWithDynamicGraph<StructureDefinitionId> {

  @CommandHandler
  @CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
  public void handle(
      ImportStructureDefinition command,
      @Autowired DynamicCommandProcessor commandProcessor
  ) {
    this.commandProcessor = commandProcessor;
    this.processCommandDynamically(
        new FixedImportStructureDefinition(
            command.getTargetIdentifier(),
            StructureDefinitionNormalizer.normalize(command.getStructureDefinitionSource())
        ),
        MissingFieldResolvingStrategy.LENIENT
    );
  }

  @EventSourcingHandler
  public void on(StructureDefinitionImported event) {
    this.onAggregateCreated(event);
  }
}