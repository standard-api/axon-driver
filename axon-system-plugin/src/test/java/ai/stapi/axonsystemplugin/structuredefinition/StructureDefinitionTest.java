package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.test.domain.DomainTestCase;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.io.IOException;
import java.util.Map;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class StructureDefinitionTest extends DomainTestCase {

  @Autowired
  private ObjectMapper objectMapper;

  @Test
  void itCanImportStructureDefinitionFromSource() throws IOException {
    var fixtureFilePath = this.getFixtureFilePath("element.profile.canonical.json");
    var file = new File(fixtureFilePath);
    var fileContent = FileUtils.readFileToString(file, "UTF-8");
    Map<String, Object> structureDefinitionSource;
    try {
      structureDefinitionSource = this.objectMapper.readValue(
          fileContent,
          new TypeReference<>() {
          }
      );
    } catch (JsonProcessingException e) {
      throw new RuntimeException(e);
    }
    var command = new ImportStructureDefinition(
        new StructureDefinitionId("Element"),
        structureDefinitionSource
    );
    this.whenCommandIsDispatched(command);
    this.thenLastEventOfTypeGraphApproved(StructureDefinitionImported.class);
  }
}