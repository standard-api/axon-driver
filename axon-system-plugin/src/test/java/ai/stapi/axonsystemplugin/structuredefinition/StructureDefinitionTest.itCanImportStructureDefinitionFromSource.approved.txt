nodes:
	node_type: ElementDefinition
	node_hash: 4DCE1
	node_edges:
		ElementDefinition -> mapping -> ElementDefinitionMapping (4DCE1 -> 18167 -> 928F3)
		ElementDefinition -> slicing -> ElementDefinitionSlicing (4DCE1 -> 6E4A9 -> 3F60D)
		ElementDefinition -> type -> ElementDefinitionType (4DCE1 -> AB28A -> E57DB)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 0DAF7 -> 4DCE1)
	attributes:
		alias -> [extensions, user content]
		comment -> There can be no stigma associated with the use of extensions by any application, project, or standard - regardless of the institution or jurisdiction that uses or defines the extensions.  The use of extensions is what allows the FHIR specification to retain a core level of simplicity for everyone.
		definition -> May be used to represent additional information that is not part of the basic definition of the element. To make the use of extensions safe and manageable, there is a strict set of governance  applied to the definition and use of extensions. Though any implementer can define an extension, there is a set of requirements that SHALL be met as part of the definition of the extension.
		max -> *
		min -> 0
		path -> Element.extension
		short -> Additional content defined by implementations

	node_type: ElementDefinition
	node_hash: B0098
	node_edges:
		ElementDefinition -> mapping -> ElementDefinitionMapping (B0098 -> C3087 -> 78C8B)
		ElementDefinition -> type -> ElementDefinitionType (B0098 -> C0E23 -> B083E)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 22A38 -> B0098)
	attributes:
		definition -> Unique id for the element within a resource (for internal references). This may be any string value that does not contain spaces.
		max -> 1
		min -> 0
		path -> Element.id
		representation -> [xmlAttr]
		short -> Unique id for inter-element referencing

	node_type: ElementDefinition
	node_hash: CA2AB
	node_edges:
		ElementDefinition -> constraint -> ElementDefinitionConstraint (CA2AB -> 21042 -> 4D49D)
		ElementDefinition -> extension -> Extension (CA2AB -> 20875 -> 9658B)
		ElementDefinition -> extension -> Extension (CA2AB -> 7E222 -> 95301)
		ElementDefinition -> mapping -> ElementDefinitionMapping (CA2AB -> 46EAE -> 83EC2)
		StructureDefinition -> differential -> StructureDefinitionDifferential (4FA2E -> 219AC -> 81CD0)
		StructureDefinition -> extension -> Extension (4FA2E -> 807B1 -> D6021)
		StructureDefinition -> extension -> Extension (4FA2E -> E6293 -> 18CF7)
		StructureDefinition -> mapping -> StructureDefinitionMapping (4FA2E -> 17020 -> 05878)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 8CEBB -> CA2AB)
	attributes:
		condition -> [ele-1]
		definition -> Base definition for all elements in a resource.
		max -> *
		min -> 0
		path -> Element
		short -> Base for all elements

	node_type: ElementDefinitionConstraint
	node_hash: 4D49D
	node_edges:
		ElementDefinition -> constraint -> ElementDefinitionConstraint (CA2AB -> 21042 -> 4D49D)
	attributes:
		expression -> hasValue() or (children().count() > id.count()) or $this is Parameters
		human -> All FHIR elements must have a @value or children unless an empty Parameters resource
		key -> ele-1
		severity -> error
		source -> http://hl7.org/fhir/StructureDefinition/Element
		xpath -> @value|f:*|h:div|self::f:Parameters

	node_type: ElementDefinitionMapping
	node_hash: 78C8B
	node_edges:
		ElementDefinition -> mapping -> ElementDefinitionMapping (B0098 -> C3087 -> 78C8B)
	attributes:
		identity -> rim
		map -> n/a

	node_type: ElementDefinitionMapping
	node_hash: 83EC2
	node_edges:
		ElementDefinition -> mapping -> ElementDefinitionMapping (CA2AB -> 46EAE -> 83EC2)
	attributes:
		identity -> rim
		map -> n/a

	node_type: ElementDefinitionMapping
	node_hash: 928F3
	node_edges:
		ElementDefinition -> mapping -> ElementDefinitionMapping (4DCE1 -> 18167 -> 928F3)
	attributes:
		identity -> rim
		map -> n/a

	node_type: ElementDefinitionSlicing
	node_hash: 3F60D
	node_edges:
		ElementDefinition -> slicing -> ElementDefinitionSlicing (4DCE1 -> 6E4A9 -> 3F60D)
		ElementDefinitionSlicing -> discriminator -> ElementDefinitionSlicingDiscriminator (3F60D -> 7C0EA -> 60B3C)
	attributes:
		description -> Extensions are always sliced by (at least) url
		rules -> open

	node_type: ElementDefinitionSlicingDiscriminator
	node_hash: 60B3C
	node_edges:
		ElementDefinitionSlicing -> discriminator -> ElementDefinitionSlicingDiscriminator (3F60D -> 7C0EA -> 60B3C)
	attributes:
		path -> url
		type -> value

	node_type: ElementDefinitionType
	node_hash: B083E
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (B0098 -> C0E23 -> B083E)
		ElementDefinitionType -> codeRef -> StructureDefinition (B083E -> B32E1 -> 951DC)
		ElementDefinitionType -> extension -> Extension (B083E -> 185EA -> F3CAF)
	attributes:
		code -> http://hl7.org/fhirpath/System.String

	node_type: ElementDefinitionType
	node_hash: E57DB
	node_edges:
		ElementDefinition -> type -> ElementDefinitionType (4DCE1 -> AB28A -> E57DB)
		ElementDefinitionType -> codeRef -> StructureDefinition (E57DB -> 90618 -> C7EE9)
	attributes:
		code -> Extension

	node_type: Extension
	node_hash: 18CF7
	node_edges:
		StructureDefinition -> extension -> Extension (4FA2E -> E6293 -> 18CF7)
	attributes:
		url -> http://hl7.org/fhir/StructureDefinition/structuredefinition-normative-version

	node_type: Extension
	node_hash: 95301
	node_edges:
		ElementDefinition -> extension -> Extension (CA2AB -> 7E222 -> 95301)
	attributes:
		url -> http://hl7.org/fhir/StructureDefinition/structuredefinition-normative-version

	node_type: Extension
	node_hash: 9658B
	node_edges:
		ElementDefinition -> extension -> Extension (CA2AB -> 20875 -> 9658B)
	attributes:
		url -> http://hl7.org/fhir/StructureDefinition/structuredefinition-standards-status

	node_type: Extension
	node_hash: D6021
	node_edges:
		StructureDefinition -> extension -> Extension (4FA2E -> 807B1 -> D6021)
	attributes:
		url -> http://hl7.org/fhir/StructureDefinition/structuredefinition-standards-status

	node_type: Extension
	node_hash: F3CAF
	node_edges:
		ElementDefinitionType -> extension -> Extension (B083E -> 185EA -> F3CAF)
	attributes:
		url -> http://hl7.org/fhir/StructureDefinition/structuredefinition-fhir-type

	node_type: StructureDefinition
	node_hash: 4FA2E
	node_edges:
		ElementDefinition -> constraint -> ElementDefinitionConstraint (CA2AB -> 21042 -> 4D49D)
		ElementDefinition -> extension -> Extension (CA2AB -> 20875 -> 9658B)
		ElementDefinition -> extension -> Extension (CA2AB -> 7E222 -> 95301)
		ElementDefinition -> mapping -> ElementDefinitionMapping (CA2AB -> 46EAE -> 83EC2)
		StructureDefinition -> differential -> StructureDefinitionDifferential (4FA2E -> 219AC -> 81CD0)
		StructureDefinition -> extension -> Extension (4FA2E -> 807B1 -> D6021)
		StructureDefinition -> extension -> Extension (4FA2E -> E6293 -> 18CF7)
		StructureDefinition -> mapping -> StructureDefinitionMapping (4FA2E -> 17020 -> 05878)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 8CEBB -> CA2AB)
	attributes:
		abstract -> true
		date -> 2022-05-28T02:47:40Z
		description -> Base StructureDefinition for Element Type: Base definition for all elements in a resource.
		experimental -> false
		fhirVersion -> 4.3.0
		kind -> complex-type
		name -> Element
		publisher -> HL7 FHIR Standard
		status -> active
		type -> Element
		url -> http://hl7.org/fhir/StructureDefinition/Element
		version -> 4.3.0

	node_type: StructureDefinition
	node_hash: 951DC
	node_edges:
		ElementDefinitionType -> codeRef -> StructureDefinition (B083E -> B32E1 -> 951DC)
	attributes:

	node_type: StructureDefinition
	node_hash: C7EE9
	node_edges:
		ElementDefinitionType -> codeRef -> StructureDefinition (E57DB -> 90618 -> C7EE9)
	attributes:

	node_type: StructureDefinitionDifferential
	node_hash: 81CD0
	node_edges:
		StructureDefinition -> differential -> StructureDefinitionDifferential (4FA2E -> 219AC -> 81CD0)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 0DAF7 -> 4DCE1)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 22A38 -> B0098)
		StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 8CEBB -> CA2AB)
	attributes:

	node_type: StructureDefinitionMapping
	node_hash: 05878
	node_edges:
		StructureDefinition -> mapping -> StructureDefinitionMapping (4FA2E -> 17020 -> 05878)
	attributes:
		identity -> rim
		name -> RIM Mapping
		uri -> http://hl7.org/v3

edges:
	edge_type: codeRef
	edge_hash: 90618
	ElementDefinitionType -> codeRef -> StructureDefinition (E57DB -> 90618 -> C7EE9)
	attributes:

	edge_type: codeRef
	edge_hash: B32E1
	ElementDefinitionType -> codeRef -> StructureDefinition (B083E -> B32E1 -> 951DC)
	attributes:

	edge_type: constraint
	edge_hash: 21042
	ElementDefinition -> constraint -> ElementDefinitionConstraint (CA2AB -> 21042 -> 4D49D)
	attributes:

	edge_type: differential
	edge_hash: 219AC
	StructureDefinition -> differential -> StructureDefinitionDifferential (4FA2E -> 219AC -> 81CD0)
	attributes:

	edge_type: discriminator
	edge_hash: 7C0EA
	ElementDefinitionSlicing -> discriminator -> ElementDefinitionSlicingDiscriminator (3F60D -> 7C0EA -> 60B3C)
	attributes:

	edge_type: element
	edge_hash: 0DAF7
	StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 0DAF7 -> 4DCE1)
	attributes:

	edge_type: element
	edge_hash: 22A38
	StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 22A38 -> B0098)
	attributes:

	edge_type: element
	edge_hash: 8CEBB
	StructureDefinitionDifferential -> element -> ElementDefinition (81CD0 -> 8CEBB -> CA2AB)
	attributes:

	edge_type: extension
	edge_hash: 185EA
	ElementDefinitionType -> extension -> Extension (B083E -> 185EA -> F3CAF)
	attributes:

	edge_type: extension
	edge_hash: 20875
	ElementDefinition -> extension -> Extension (CA2AB -> 20875 -> 9658B)
	attributes:

	edge_type: extension
	edge_hash: 7E222
	ElementDefinition -> extension -> Extension (CA2AB -> 7E222 -> 95301)
	attributes:

	edge_type: extension
	edge_hash: 807B1
	StructureDefinition -> extension -> Extension (4FA2E -> 807B1 -> D6021)
	attributes:

	edge_type: extension
	edge_hash: E6293
	StructureDefinition -> extension -> Extension (4FA2E -> E6293 -> 18CF7)
	attributes:

	edge_type: mapping
	edge_hash: 17020
	StructureDefinition -> mapping -> StructureDefinitionMapping (4FA2E -> 17020 -> 05878)
	attributes:

	edge_type: mapping
	edge_hash: 18167
	ElementDefinition -> mapping -> ElementDefinitionMapping (4DCE1 -> 18167 -> 928F3)
	attributes:

	edge_type: mapping
	edge_hash: 46EAE
	ElementDefinition -> mapping -> ElementDefinitionMapping (CA2AB -> 46EAE -> 83EC2)
	attributes:

	edge_type: mapping
	edge_hash: C3087
	ElementDefinition -> mapping -> ElementDefinitionMapping (B0098 -> C3087 -> 78C8B)
	attributes:

	edge_type: slicing
	edge_hash: 6E4A9
	ElementDefinition -> slicing -> ElementDefinitionSlicing (4DCE1 -> 6E4A9 -> 3F60D)
	attributes:

	edge_type: type
	edge_hash: AB28A
	ElementDefinition -> type -> ElementDefinitionType (4DCE1 -> AB28A -> E57DB)
	attributes:

	edge_type: type
	edge_hash: C0E23
	ElementDefinition -> type -> ElementDefinitionType (B0098 -> C0E23 -> B083E)
	attributes:
